adr/0003-PlacementNew.md

```diff
@@ -0,0 +1,110 @@
+# C++ Data Lifetime Management via Placement New
+
+## Status
+PROPOSED
+
+Proposed by: Carlos Reyes (February 18, 2022)
+
+Discussed with:
+
+* Benson Ma (early ideas)
+* Eduardo Gonzalez (overall design goals)
+
+## Context
+Our architecture has a low-level C/C++ layer and a high level Java/Scala layer. We use JavaCpp to communicate between
+the layers. We would like to only use C++ code in the low level layer.
+
+Right now we allocate return values using C++ and deallocate them from Scala.
+This requires us to use C-style calloc/free system calls. We would like to migrate away from this design.
```

Comment:
JavaCPP binds to C++ so it's not the reason we have to use C-style APIs here.  The reason is AVEO is a C API and we have to use that to run things on the VE.  Because AVEO is a RPC system, we can't use new/delete directly.  The fact that the memory lives in this other process/device is the main reason things are complicated.